sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1135,17 +1135,18 @@ class DataFrame private[sql](
    * @group dfops
    * @since 1.3.0
    */
-  def withColumn(colName: String, col: Column): DataFrame = {
+  def withColumn(colName: String, col: Column, metadata: Option[Metadata] = None): DataFrame = {
```

Comment:
Can you do this in a different PR?  Also we should do it without using `Option` and default arguments so that it works well in Java.